In [1]:
import pandas as pd
import numpy as np

In [2]:
enrollment = pd.read_excel('Performance_Task.xlsx',sheet_name = 'Enrollment')

In [3]:
enrollment.head()

student_id school_name  grade_level_2019
0       13898    School G                 1
1       27795    School G                 1
2       22938    School E                 1
3       22431    School F                 1
4       18048    School E                 1

In [4]:
enrollment.shape

(2788, 3)

In [5]:
MAP_scores = pd.read_excel('Performance_Task.xlsx',sheet_name = 'MAP Scores')

In [6]:
MAP_scores.head()

student_id  year map_term  subject rit_score percentile quartile  \
0       13898  2019     Fall  Reading       161         59        3   
1       27795  2019     Fall  Reading       158         50        3   
2       22938  2019     Fall  Reading       162         62        3   
3       22431  2019     Fall  Reading       162         62        3   
4       18048  2019     Fall  Reading       166         73        3   

  typical_fall_to_spring_growth  
0                            18  
1                            19  
2                            18  
3                            18  
4                            17

In [7]:
MAP_scores.shape

(11118, 8)

In [8]:
data = pd.merge(enrollment, MAP_scores, on = 'student_id')

In [9]:
data.head()

student_id school_name  grade_level_2019  year map_term  subject rit_score  \
0       13898    School G                 1  2019     Fall  Reading       161   
1       13898    School G                 1  2019     Fall     Math       166   
2       13898    School G                 1  2019   Spring  Reading       175   
3       13898    School G                 1  2019   Spring     Math       186   
4       27795    School G                 1  2019     Fall  Reading       158   

  percentile quartile typical_fall_to_spring_growth  
0         59        3                            18  
1         69        3                            20  
2         43        2                             .  
3         65        3                             .  
4         50        3                            19

In [10]:
Spec_Programs = pd.read_excel('Performance_Task.xlsx',sheet_name = 'Special Programs')

In [11]:
Spec_Programs.head()

student_id program_name
0       13898         SPED
1       27795         SPED
2       22938            .
3       22431            .
4       18048            .

In [12]:
Spec_Programs.shape

(2787, 2)

In [13]:
data = pd.merge(data,Spec_Programs, on = 'student_id')

In [14]:
data.head()

student_id school_name  grade_level_2019  year map_term  subject rit_score  \
0       13898    School G                 1  2019     Fall  Reading       161   
1       13898    School G                 1  2019     Fall     Math       166   
2       13898    School G                 1  2019   Spring  Reading       175   
3       13898    School G                 1  2019   Spring     Math       186   
4       27795    School G                 1  2019     Fall  Reading       158   

  percentile quartile typical_fall_to_spring_growth program_name  
0         59        3                            18         SPED  
1         69        3                            20         SPED  
2         43        2                             .         SPED  
3         65        3                             .         SPED  
4         50        3                            19         SPED

In [15]:
data.tail()

student_id school_name  grade_level_2019  year map_term subject  \
11117       24197    School D                 8  2019   Spring    Math   
11118       27656    School D                 8  2019     Fall    Math   
11119       27656    School D                 8  2019   Spring    Math   
11120       29777    School E                 0  2019     Fall    Math   
11121       29777    School E                 0  2019   Spring    Math   

      rit_score percentile quartile typical_fall_to_spring_growth program_name  
11117       226         40        2                             .            .  
11118       199          5        1                             6            .  
11119         .          .        .                             .            .  
11120       133         32        2                            21            .  
11121         .          .        .                             .            .

In [16]:
data.shape

(11122, 11)

All the features are needed to perform exploratory data analysis except student_id. 
student_id was needed to merge the tables together but it will not give us any valuable information.  This feature is a unique student identifier and would be valuable if individual teachers were given this information to use to identify which topics they need to reteach or spend more time on in the future.  We will keep ths feature for now. 

In [17]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 11122 entries, 0 to 11121
Data columns (total 11 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   student_id                     11122 non-null  int64 
 1   school_name                    11122 non-null  object
 2   grade_level_2019               11122 non-null  int64 
 3   year                           11122 non-null  int64 
 4   map_term                       11122 non-null  object
 5   subject                        11122 non-null  object
 6   rit_score                      11122 non-null  object
 7   percentile                     11122 non-null  object
 8   quartile                       11122 non-null  object
 9   typical_fall_to_spring_growth  11122 non-null  object
 10  program_name                   11122 non-null  object
dtypes: int64(3), object(8)
memory usage: 1.0+ MB


In [18]:
data.describe()

student_id  grade_level_2019     year
count  11122.000000      11122.000000  11122.0
mean   19926.775220          4.227837   2019.0
std     5732.455867          2.714756      0.0
min    10000.000000          0.000000   2019.0
25%    14981.000000          2.000000   2019.0
50%    19856.000000          5.000000   2019.0
75%    24873.000000          7.000000   2019.0
max    29996.000000          8.000000   2019.0

In [19]:
data['school_name'].unique()

array(['School G', 'School E', 'School F', 'School B', 'School C',
       'School D', 'School A'], dtype=object)

In [20]:
data['school_name'].value_counts()

School E    2154
School F    1724
School D    1586
School A    1482
School C    1440
School B    1422
School G    1314
Name: school_name, dtype: int64

There are 7 schools in total.  Each letter represents a school.

In [21]:
data['grade_level_2019'].unique()

array([1, 2, 3, 4, 5, 6, 7, 8, 0])

In [22]:
data['grade_level_2019'].value_counts()

6    1574
7    1488
8    1476
5    1392
2    1280
1    1278
0    1270
3     900
4     464
Name: grade_level_2019, dtype: int64

This data is for students from Pre-K to 8th grade.  Pre-K to 4th is primary school.  Middle school is 5th to 8th.  

The number of student scores for the MAP test decreased tremedously from 2nd grade to 3rd and then to 4th.  The number of scores increased by 928 from 4th to 5th grade. 

Lets group by school and grade if there was a school that did not submit scores for a particuar school 

In [23]:
Middle_School = data[data['grade_level_2019'] > 4] 

In [24]:
Middle_School['school_name'].unique()

array(['School B', 'School C', 'School D', 'School A'], dtype=object)

In [25]:
Primary_School = data[data['grade_level_2019'] < 5]

In [26]:
pd.pivot_table(Primary_School, index = ['grade_level_2019'], columns = ['school_name'], values = 'student_id', aggfunc = 'count')



school_name       School E  School F  School G
grade_level_2019                              
0                    402.0     428.0     440.0
1                    392.0     436.0     450.0
2                    432.0     424.0     424.0
3                    464.0     436.0       NaN
4                    464.0       NaN       NaN

The above table shows that School F did not submit MAP scores for 4th grade and School G did not submit MAP scores for 3rd and 4th grade.  This is interesting but now we know why the total amount is drastically lower between 2nd and 3rd, 3rd and 4th, and 4th and 5th.  We could contact the school to see why this information was not submitted but for this data analysis we will continue with what was given to us. 

In [27]:
data['year'].unique()

array([2019])

In [28]:
data['year'].value_counts()

2019    11122
Name: year, dtype: int64

All rows have 2019 as the year. This is perfect! 

In [29]:
data['map_term'].unique()

array(['Fall', 'Spring'], dtype=object)

In [30]:
data['map_term'].value_counts()

Fall      5561
Spring    5561
Name: map_term, dtype: int64

In [31]:
data['subject'].unique()

array(['Reading', 'Math'], dtype=object)

In [32]:
data['subject'].value_counts()

Math       5562
Reading    5560
Name: subject, dtype: int64

In [33]:
data['student_id'].value_counts()

11220    8
20482    4
19028    4
19036    4
25185    4
        ..
16192    2
24197    2
19113    2
11654    2
27152    2
Name: student_id, Length: 2787, dtype: int64

In [34]:
data[data['student_id'] == 11220]

student_id school_name  grade_level_2019  year map_term  subject  \
6830       11220    School D                 6  2019     Fall  Reading   
6831       11220    School D                 6  2019     Fall     Math   
6832       11220    School D                 6  2019   Spring  Reading   
6833       11220    School D                 6  2019   Spring     Math   
6834       11220    School D                 5  2019     Fall  Reading   
6835       11220    School D                 5  2019     Fall     Math   
6836       11220    School D                 5  2019   Spring  Reading   
6837       11220    School D                 5  2019   Spring     Math   

     rit_score percentile quartile typical_fall_to_spring_growth program_name  
6830         .          .        .                             .            .  
6831       197         16        1                            10            .  
6832       215         48        2                             .            .  
6833       202          8        1                             .            .  
6834         .          .        .                             .            .  
6835       197         16        1                            10            .  
6836       215         48        2                             .            .  
6837       202          8        1                             .            .

Student 11220 is a duplicate.  This student has a grade level of 5th and 6th.  Since we do not have any prior data to determine the actual grade for the this student we will remove this student from the data set instead of guessing which grade level the student is in. Guessing the grade could produce inaccurate data analysis.  

In [35]:
#Drop duplicate

data.drop(data[data['student_id'] == 11220].index, inplace = True)

In [36]:
Fall_scores = data[data['map_term'] == 'Fall']
Spring_scores = data[data['map_term'] == 'Spring']

In [37]:
Fall_scores.shape

(5557, 11)

In [38]:
Spring_scores.shape

(5557, 11)

In [39]:
data['rit_score'].unique()

array([161, 166, 175, 186, 158, 135, 174, 182, 162, 160, 178, 180, 164,
       151, 181, 163, '.', 159, 167, 153, 157, 183, 176, 192, 204, 165,
       171, 188, 177, 190, 196, 202, 185, 222, 173, 156, 201, 203, 168,
       184, 187, 189, 194, 172, 169, 150, 199, 140, 170, 218, 155, 147,
       179, 191, 193, 198, 148, 145, 125, 143, 195, 212, 123, 126, 149,
       152, 197, 209, 142, 128, 208, 213, 200, 133, 215, 154, 144, 137,
       207, 206, 146, 138, 205, 217, 139, 223, 220, 216, 227, 136, 127,
       210, 224, 141, 214, 211, 221, 130, 226, 219, 228, 229, 225, 240,
       235, 234, 230, 236, 232, 237, 242, 243, 231, 244, 238, 233, 239,
       250, 241, 252, 246, 248, 249, 245, 255, 253, 251, 247, 258, 257,
       256, 262, 259, 269, 265, 261, 266, 267, 260, 254, 129, 134, 120,
       114, 132, 131, 124, 109, 118, 116, 119, 122, 121, 115, 104, 113,
       107, 110, 117], dtype=object)

In [40]:
data['rit_score'].value_counts()

.      997
221    180
210    163
209    162
212    159
      ... 
258      1
110      1
117      1
113      1
104      1
Name: rit_score, Length: 159, dtype: int64

We noticed that there are 997 spaces with no rit_score.  Also, each value is an object instead of a integer.  We need to do two things.

1. Replace the '.' with an empty space
2. convert the object to an integer

In [41]:
#We do not need this.  This can be dropped.

data = data.replace('.',' ')

In [42]:
data['rit_score'].value_counts()

       997
221    180
210    163
209    162
212    159
      ... 
258      1
110      1
117      1
113      1
104      1
Name: rit_score, Length: 159, dtype: int64

In [43]:
data[data['rit_score'] == ' ']

student_id school_name  grade_level_2019  year map_term  subject  \
20          20392    School G                 1  2019     Fall  Reading   
21          20392    School G                 1  2019     Fall     Math   
63          26626    School F                 1  2019   Spring     Math   
144         11466    School G                 1  2019     Fall  Reading   
145         11466    School G                 1  2019     Fall     Math   
...           ...         ...               ...   ...      ...      ...   
11113       28744    School D                 7  2019   Spring     Math   
11115       12169    School A                 7  2019   Spring     Math   
11116       24197    School D                 8  2019     Fall     Math   
11119       27656    School D                 8  2019   Spring     Math   
11121       29777    School E                 0  2019   Spring     Math   

      rit_score percentile quartile typical_fall_to_spring_growth program_name  
20                                                                        SPED  
21                                                                        SPED  
63                                                                        SPED  
144                                                                             
145                                                                             
...         ...        ...      ...                           ...          ...  
11113                                                                           
11115                                                                           
11116                                                                           
11119                                                                           
11121                                                                           

[997 rows x 11 columns]

In [44]:
#Drop each row that does not have a rit_score.  If there is not rit_score then there is not data to be used

data.drop(data[data['rit_score'] == ' '].index, inplace = True)

In [45]:
data.shape

(10117, 11)

In [46]:
data['rit_score'] = data['rit_score'].astype(str).astype(int)

In [47]:
data['rit_score'].dtypes

dtype('int64')

In [48]:
data['percentile'].unique()

array([59, 69, 43, 65, 50, 4, 40, 53, 62, 52, 51, 48, 64, 73, 28, 10, 6,
       29, 72, 5, 97, 89, 84, 96, 70, 81, 77, 58, 46, 56, 68, 90, 87, 34,
       60, 36, 86, 94, 55, 99, 38, 39, 41, 8, 24, 95, 78, 74, 67, 79, 83,
       80, 26, 33, 31, 93, 76, 47, 19, 91, 49, 16, 21, 25, 7, 44, 30, 45,
       98, 82, 88, 54, 35, 15, 1, 2, 12, 14, 92, 75, 17, 23, 11, 3, 32,
       42, 85, 61, 57, 37, 13, 9, 18, 66, 22, 27, 20, 71, 63],
      dtype=object)

In [49]:
data['percentile'].value_counts()

56    187
1     186
35    182
40    179
64    178
     ... 
11     54
75     53
34     45
36     44
63     40
Name: percentile, Length: 99, dtype: int64

In [50]:
data['percentile'] = data['percentile'].astype(str).astype(int)

In [51]:
data['quartile'].unique()

array([3, 2, 1, 4], dtype=object)

In [52]:
data['quartile'].value_counts()

3    3007
2    2696
4    2237
1    2177
Name: quartile, dtype: int64

In [53]:
data['quartile'] = data['quartile'].astype(str).astype(int)

In [54]:
data['quartile'].dtypes

dtype('int64')

In [55]:
data['typical_fall_to_spring_growth'].unique()

array([18, 20, ' ', 19, 27, 21, 17, 23, 13, 14, 22, 12, 16, 15, 26, 24,
       11, 30, 25, 29, 28, 9, 10, 6, 7, 5, 8, 4, 3, 2], dtype=object)

In [56]:
data['typical_fall_to_spring_growth'].value_counts()

      5046
10     561
4      480
5      480
2      384
3      349
6      328
7      257
11     213
18     206
8      199
12     196
9      181
20     170
13     156
19     144
17     121
16     118
21     117
15      96
14      86
22      85
23      69
24      32
25      16
26      13
27       8
30       3
29       2
28       1
Name: typical_fall_to_spring_growth, dtype: int64

In [66]:
#Replace ' ' with 0 then change the data from a string to a number

data['typical_fall_to_spring_growth'] = data['typical_fall_to_spring_growth'].replace(' ', 0)

data['typical_fall_to_spring_growth'] = data['typical_fall_to_spring_growth'].astype(str).astype(int)

Some data is missing here but we will not delete those rows.  Students may have taken the MAP test in the Fall and not the Spring or vice versa.  We will make a separate dataframe with no empty cells and call it data_perfect.

In [67]:
data_perfect = data.copy()
data_perfect.drop(data_perfect[data_perfect['typical_fall_to_spring_growth'] == ' '].index, inplace = True)

In [68]:
data['program_name'].unique()

array(['SPED', ' ', 'LEP', 'Tier 2', 'Gifted', 504], dtype=object)

In [69]:
#Replace ' ' with General for General Student

data['program_name'] = data['program_name'].replace(' ', 'General')
data_perfect['program_name'] = data_perfect['program_name'].replace(' ', 'General')

In [70]:
data['program_name'].unique()

array(['SPED', 'General', 'LEP', 'Tier 2', 'Gifted', 504], dtype=object)

In [71]:
data['program_name'].value_counts()

General    8157
Gifted      734
SPED        698
Tier 2      331
LEP         122
504          75
Name: program_name, dtype: int64

General: Student without special education ability 

Gifted: Student is able to perofrm at a higher level

SPED: Student is provided accomodations to assist with his/her education

Tier 2: Student needs small group instruction

LEP: Student has limit english proficiency

504: Student has a disiblity under the law (ex. asthma)

In [72]:
data.head()

student_id school_name  grade_level_2019  year map_term  subject  \
0       13898    School G                 1  2019     Fall  Reading   
1       13898    School G                 1  2019     Fall     Math   
2       13898    School G                 1  2019   Spring  Reading   
3       13898    School G                 1  2019   Spring     Math   
4       27795    School G                 1  2019     Fall  Reading   

   rit_score  percentile  quartile  typical_fall_to_spring_growth program_name  
0        161          59         3                             18         SPED  
1        166          69         3                             20         SPED  
2        175          43         2                              0         SPED  
3        186          65         3                              0         SPED  
4        158          50         3                             19         SPED

In [73]:
data.shape

(10117, 11)

In [74]:
data_perfect.head()

student_id school_name  grade_level_2019  year map_term  subject  \
0       13898    School G                 1  2019     Fall  Reading   
1       13898    School G                 1  2019     Fall     Math   
2       13898    School G                 1  2019   Spring  Reading   
3       13898    School G                 1  2019   Spring     Math   
4       27795    School G                 1  2019     Fall  Reading   

   rit_score  percentile  quartile  typical_fall_to_spring_growth program_name  
0        161          59         3                             18         SPED  
1        166          69         3                             20         SPED  
2        175          43         2                              0         SPED  
3        186          65         3                              0         SPED  
4        158          50         3                             19         SPED

In [75]:
data_perfect.shape

(10117, 11)

In [76]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10117 entries, 0 to 11120
Data columns (total 11 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   student_id                     10117 non-null  int64 
 1   school_name                    10117 non-null  object
 2   grade_level_2019               10117 non-null  int64 
 3   year                           10117 non-null  int64 
 4   map_term                       10117 non-null  object
 5   subject                        10117 non-null  object
 6   rit_score                      10117 non-null  int64 
 7   percentile                     10117 non-null  int64 
 8   quartile                       10117 non-null  int64 
 9   typical_fall_to_spring_growth  10117 non-null  int64 
 10  program_name                   10117 non-null  object
dtypes: int64(7), object(4)
memory usage: 1.2+ MB
